sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ClassAndMethodName.cs

```diff
@@ -0,0 +1,341 @@
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using System.Collections.Immutable;
+using System.Text;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId_TypeName)]
+    [Rule(DiagnosticId_MethodName)]
+    public class ClassAndMethodName : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId_MethodName = "S101";
+        internal const string DiagnosticId_TypeName = "S100";
+
+        private const string MessageFormat = "Rename {0} '{1}' to match camel case naming rules, {2}.";
+        internal const string MessageFormatNonUnderscore = "consider using '{0}'";
+        internal const string MessageFormatUnderscore = "trim underscores from the name";
+
+        private static readonly DiagnosticDescriptor rule_MethodName =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId_TypeName, MessageFormat, RspecStrings.ResourceManager);
+
+        private static readonly DiagnosticDescriptor rule_TypeName =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId_MethodName, MessageFormat, RspecStrings.ResourceManager);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => ImmutableArray.Create(rule_MethodName, rule_TypeName);
+
+        private static readonly Dictionary<SyntaxKind, string> TypeKindNameMapping = new Dictionary<SyntaxKind, string>
+        {
+            {SyntaxKind.StructDeclaration, "struct" },
+            {SyntaxKind.ClassDeclaration, "class" },
+            {SyntaxKind.InterfaceDeclaration, "interface" },
+            {SyntaxKind.MethodDeclaration, "method" },
+            {SyntaxKind.PropertyDeclaration, "property" }
+        };
+
+        protected sealed override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c => CheckTypeName((BaseTypeDeclarationSyntax)c.Node, c),
+                SyntaxKind.ClassDeclaration,
+                SyntaxKind.InterfaceDeclaration,
+                SyntaxKind.StructDeclaration);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var declaration = (MethodDeclarationSyntax)c.Node;
+                    CheckMemberName(declaration, declaration.Identifier, c);
+                },
+                SyntaxKind.MethodDeclaration);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var declaration = (PropertyDeclarationSyntax)c.Node;
+                    CheckMemberName(declaration, declaration.Identifier, c);
+                },
+                SyntaxKind.PropertyDeclaration);
+        }
+
+        private static void CheckTypeName(BaseTypeDeclarationSyntax type,
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this unused method parameter 'type'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1172)